### PR TITLE
Generic views create edit delete meetings

### PR DESCRIPTION
**Pull request by nuttapol-kor**
- User generic views to create, edit, delete meetings.
- Edit, delete, kick feature.

### CODE REVIEW

File: appointment/forms.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from django import forms
+from .models import Meeting
+
+
+class DateInput(forms.DateInput):
+    """Datetime picker type."""
+    input_type = 'datetime-local'
```
Comment (nuttapol-kor) [inline appointment/forms.py:7]: Use for use date-picker in start-time, end-time. ( firefox, safari not supported )

File: appointment/models.py [hunk 2]
```diff
@@ -15,7 +15,8 @@ class Meeting(models.Model):
     end_time = models.DateTimeField()
     location = models.CharField(max_length=100)
     contact = models.CharField(max_length=100)
-    upload = models.FileField(upload_to='doc/pdfs')
+    upload = models.FileField(upload_to='doc/pdfs', blank=True, null=True)
```
Comment (nuttapol-kor) [inline appointment/models.py:18]: not required to upload a flie

File: appointment/models.py [hunk 3]
```diff
@@ -15,7 +15,8 @@ class Meeting(models.Model):
...
+    host = models.ForeignKey(User, on_delete=models.CASCADE, null=True)
```
Comment (nuttapol-kor) [inline appointment/models.py:19]: the meeting has a host user now

File: appointment/templates/appointment/base.html [hunk 4]
```diff
@@ -117,7 +117,7 @@
                         <a class="nav-link" href="{% url 'profile' %}">Profile</a>
                     </li>
                     <li class="nav-item">
-                        <a class="nav-link" href="{% url 'create' %}">Create meeting</a>
+                        <a class="nav-link" href="{% url 'appointment:create-meeting' %}">Create meeting</a>
```
Comment (nuttapol-kor) [inline appointment/templates/appointment/base.html]: Use generic view create-meeting instead.

File: appointment/templates/appointment/base.html [hunk 5]
```diff
@@ -166,7 +166,7 @@ <h4 style="font-size: 25px; color: aliceblue">{{ request.user }} </h4>
                     {% if user.is_authenticated %}
                         <li><a href="{% url 'profile' %}"><i class="fa fa-user-circle"></i>Profile</a></li>
                         <li><a href="#"><i class="fa fa-table"></i>Subscriptions</a></li>
-                        <li><a href="{% url 'create' %}"><i class="fa fa-plus-square"></i>Create meeting</a></li>
+                        <li><a href="{% url 'appointment:create-meeting' %}"><i class="fa fa-plus-square"></i>Create meeting</a></li>
```
Comment (nuttapol-kor) [inline appointment/templates/appointment/base.html:169]: I change to use new create-meeting instead.

File: appointment/templates/appointment/detail.html [hunk 6]
```diff
@@ -13,18 +13,17 @@ <h2>Date : {{ meeting.date_meeting }}</h2>
...
+                    <a class="btn btn-outline-primary" href="{% url 'appointment:join' meeting.id %}">Join</a>
+                    <a class="btn btn-outline-danger" href="{% url 'appointment:leave' meeting.id %}">Leave</a>
+                {% endif %}
+                <a class="btn btn-outline-info" href="{% url 'appointment:participants' meeting.id %}">Participants</a>
+                {% if meeting.host == user %}
+                <div>
+                    <a class="btn btn-primary" href="{% url 'appointment:edit-meeting' meeting.id %}">Edit</a>
+                    <a class="btn btn-danger" href="{% url 'appointment:delete-meeting' meeting.id %}">Delete</a>
+                </div>
+                {% endif %}
```
Comment (nuttapol-kor) [inline appointment/templates/appointment/detail.html:26]: use `<a>` tag instead `<form>` tag

File: appointment/templates/appointment/participants.html [hunk 7]
```diff
@@ -4,9 +4,12 @@
...
     {% if participants %}
-        {% for user in participants %}
-            <a href="{% url 'other_profiles' user.user.id %}" target="_blank"> <img src="{{ user.user.profile.image.url }}" width="50px;"></a>
-            <a href="{% url 'other_profiles' user.user.id %}" target="_blank"> {{ user.user.username }}</a>
+        {% for each_user in participants %}
+            <a href="{% url 'other_profiles' each_user.user.id %}" target="_blank"> <img src="{{ each_user.user.profile.image.url }}" width="50px;"></a>
+            <a href="{% url 'other_profiles' each_user.user.id %}" target="_blank"> {{ each_user.user.username }}</a>
+            {% if meeting.host == user and each_user.user != user %}
+                <a class="btn btn-danger" href="{% url 'appointment:kick' meeting.id each_user.user.id %}">Kick</a>
+            {% endif %}
```
Comment (nuttapol-kor) [inline appointment/templates/appointment/participants.html:12]: Only the host can see a kick button.

File: appointment/urls.py [hunk 8]
```diff
@@ -7,10 +7,14 @@
...
     path("meeting/<int:year>/<int:month>/<int:day>", views.meeting_list, name="meet_list"),
-    path("<int:meeting_id>/detail", views.detail, name="detail"),
+    path("meeting/<int:meeting_id>/detail", views.detail, name="detail"),
     path('result/', views.search, name="search"),
-    path('<int:meeting_id>/join', views.join, name='join'),
-    path('<int:meeting_id>/leave', views.leave, name='leave'),
-    path('<int:meeting_id>/participants', views.appointment_participants, name='participants'),
+    path('meeting/<int:meeting_id>/join', views.join, name='join'),
+    path('meeting/<int:meeting_id>/leave', views.leave, name='leave'),
+    path('meeting/<int:meeting_id>/participants', views.appointment_participants, name='participants'),
```
Comment (nuttapol-kor) [inline appointment/urls.py:14]: show meeting path when do something about meeting.

File: appointment/views.py [hunk 9]
```diff
@@ -145,4 +150,98 @@ def appointment_participants(request, meeting_id):
...
+    """For create a meeting"""
+    model = Meeting
+    form_class = UserCreateMeetForm
+    template_name = 'appointment/create_meeting.html'
+    success_message = "%(subject)s was created successfully!!"
+
+    def form_valid(self, form):
+        """Add host permission to user that create meeting."""
+        form.instance.host = self.request.user
+        return super().form_valid(form)
```
Comment (nuttapol-kor) [inline appointment/views.py:166]: When the user creates a meeting. That user becomes a  host in this meeting.

File: appointment/views.py [hunk 10]
```diff
@@ -145,4 +150,98 @@ def appointment_participants(request, meeting_id):
...
+
+    def test_func(self):
+        """Test the user to edit meeting. The host can editable only."""
+        meeting = self.get_object()
+        if self.request.user == meeting.host:
+            return True
+        return False
```
Comment (nuttapol-kor) [inline appointment/views.py:193]: Test the user permission (Is the host user?). The host user can edit meeting only

File: appointment/views.py [hunk 11]
```diff
@@ -145,4 +150,98 @@ def appointment_participants(request, meeting_id):
...
+    template_name = 'appointment/delete_meeting.html'
+    success_url = '/'
+    success_message = "%(subject)s was deleted successfully!!"
+
+    def test_func(self):
+        """Test the user to delete meeting. The host can deleteable only."""
+        meeting = self.get_object()
+        if self.request.user == meeting.host:
+            return True
+        return False
```
Comment (nuttapol-kor) [inline appointment/views.py:215]: Test the user permission (Is the host user?). The host user can delete meeting only.

File: tewma/urls.py [hunk 12]
```diff
@@ -28,6 +28,5 @@
     path("profile/edit_profile/", user_views.edit_profile, name='edit_profile'),
     path('login/', user_views.LoginFormView.as_view(redirect_authenticated_user=True), name='login'),
     path('logout/', auth_views.LogoutView.as_view(template_name='users/logout.html'), name='logout'),
-    path('create/', user_views.create_meet, name='create'),
```
Comment (nuttapol-kor) [inline tewma/urls.py:31]: remove old create meeting path

File: users/views.py [hunk 13]
```diff
@@ -57,33 +57,19 @@ def edit_profile(request):
...
-        if form.is_valid():
-            form.save()
-            subject = form.cleaned_data.get('subject')
-            start_time = form.cleaned_data.get('start_time').strftime('%d %B %Y')
-            messages.success(request, f'{subject} start at {start_time} has been created!!')
-            return redirect('appointment:home_page')
-    else:
-        form = UserCreateMeetForm()
-    return render(request, 'users/create_meeting.html', {'form': form})
-
```
Comment (nuttapol-kor) [inline users/views.py:74]: remove old create meeting function.

**Review by Bouncyyahomie — APPROVED** [review]
Nice design models and good new feature

**Review by Chayapol-c — APPROVED** [review]
Good idea to give a user who owns a meeting to manage his meeting but think it could improve styled.

**Pull request merged**